Add of_address classmethod to EarthLocation

adrn opened the pull request:
In the same vein as `SkyCoord.from_name()`, this PR implements a `from_address()` classmethod on `EarthLocation` that queries the Google maps API to get a latitude and longitude for a specified address. As with Google maps, the address can be a fully specified address (e.g., 123 Main St., New York, NY), or a city (e.g., Danbury), or a country, etc. etc...

cc @eteq @bmorris3

bmorris3 commented:
Awesome! I have used that API before too, and accidentally hit the query limit without intending to do a huge number of queries. Would you be open to a PR to your branch that adds a little cache, as one way to prevent users from hitting the query limit?

mhvk commented:
This looks very nice. Can the google API also return a height?

adrn commented:
There is a separate API for retrieving elevations -- I implemented an option to also query and get that as well. @bmorris3 this would require a separate cache if you still think a cache would be useful

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -235,6 +270,73 @@ def of_site(cls, site_name):
+
+        See Also
+        --------
+        https://developers.google.com/maps/documentation/geocoding/intro
```
These lines cause (I think) the sphinx build to fail. Maybe listing them as references will help? I.e.,
```
References
----------
.. [1] https://developers.google.com/maps/documentation/geocoding/intro
.. [2] https://developers.google.com/maps/documentation/elevation/intro
```

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -235,6 +270,73 @@ def of_site(cls, site_name):
+
+        This is intended as a quick convenience function to get fast access to
+        locations. In the background, this just issues a query to the Google maps
+        geocoding API. It is not meant to be abused! Google uses IP-based query
```
If using references, you could add `[1]_` here (and delete the last sentence).

mhvk commented on astropy/coordinates/earth.py:
```diff
@@ -235,6 +270,73 @@ def of_site(cls, site_name):
+            NY) or a city name (e.g., Danbury, CT), or etc.
+        get_height : bool (optional)
+            Use the retrieved location to perform a second query to the Google maps
+            elevation API to retrieve the height of the input address.
```
And add `[2]_` here.

MSeifert04 commented on astropy/coordinates/earth.py:
```diff
@@ -40,6 +46,35 @@ def _check_ellipsoid(ellipsoid=None, default='WGS84'):
+
+    results = resp_data.get('results', [])
+
+    if len(results) == 0:
```
PEP8 recommends `if not results` to check for an empty list.

mhvk commented:
@adrn - this looks great. The coverage failure likely is because you do not test the exceptions very thoroughly -- but I think this is fine since the ones you do not test are for time-out and "unknown reason".

One question: should this be `of_address` in analogy with `of_site`? Of course, one can also argue by analogy with `from_geodetic` and `from_geocentric`, but it shares with `of_site` that it is an indirect construction. cc @eteq 

Also, how will people find out about this?  But writing this makes me realize `EarthLocation` is not very much exposed at all right now -- which perhaps is not so bad as it might be sensible to refactor it somewhat (see #4261). Maybe still raise a new issue?

adrn commented:
@mhvk I'm not sure why we have both `from_` and `of_` methods -- in my own code I always use `from_` so I don't get confused -- but I switched the name to make it more like `of_site`. I also added some documentation in coordinates/index so hopefully that helps with exposure!

astrofrog commented:
If we're deciding on conventions, I agree that ``from_`` is better and more standard across packages.

adrn commented:
@astrofrog OK, do you think I should keep this as `from_address` and make a separate issue to change `of_site`  to `from_site`?

astrofrog commented:
@adrn - that's my personal opinion, but i'm not a coordinates maintainer :)

mhvk commented on docs/coordinates/index.rst:
```diff
@@ -215,16 +215,32 @@ a similar quick way to get an `~astropy.coordinates.EarthLocation`::
+Google maps, this works with fully specified addresses, location names, city
+names, and etc.::
+
+    >>> EarthLocation.of_address('1002 Holy Grail Court, St. Louis, MO')  # doctest: +REMOTE_DATA +FLOAT_CMP
```
Yay! the Holy Grail!

mhvk commented:
Writing the various forms out:
```
EarthLocation.from_geodetic(10*u.deg, 10.*u.deg)
EarthLocation.from_address('New York')
Earthlocation.of_site('keck')
Earthlocation.of_address('New York')
```
I think I start very slightly liking the `of_address` form better (and even more the `of_cite`. One way to distinguish the two might be that the `from_?` form also has (or at least in principle could have) `to_?` form, while the `of_?` does not.

mhvk commented:
Also had a quick final review: good that you found a place for it in the documentation!

Unless I hear anything from @eteq, I'll merge this by the end of the day.

mhvk commented:
Thanks, @adrn!

astrofrog commented:
> I think I start very slightly liking the of_address form better (and even more the of_site. One way to distinguish the two might be that the from_? form also has (or at least in principle could have) to_? form, while the of_? does not.

FWIW I still think ``from_`` was better because it's the 'conventional' way to name class methods...
